### PR TITLE
Lookup shards on demand with universal locator

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

This is really two endeavors that that merged into one:

1. The content claims locator, when it gets an index, fetches (in a blocking manner) locations for every single shard in the index. That means for large files or directories, you can see 20 or 30 fetches before any blocks can be retrieved. This is particularly pronounced when you're just reading a single file out of a directory. Instead, the content claims locator needs to replicate what the indexer does: just fetch the shard with the current requested digest, then fetch others as needed.
2. In looking at the indexing locator, I noticed several holes in the logic -- for the exact same scenario described above, the indexer properly returns the exact data you want on the first fetch (just the location for the shard in the index that contants the hash you looked up) but if you were to then request a hash in a different shard, it will fail. Moreover, the content claims fetcher contains several cache optimizations that have been added over time.

As I found myself adding the same logic to both fetchers, I decided to exact the common logic from the part that's specific to each service. Specifically, I decided to just write the locator for the indexer, and an adapter for the content claims service that mimics the interface of the current indexer client.

This also involved a spelunking down the stack to clean up our claim parsing. The content claims service has a well developed schema parser using ucanto's schemas, while the indexing service was using something called zod. This also moves claim schema parsing down into the js-indexer-service-client, so in this repo we can work with the same well developed schemas in the content claims repo regardless of service (this also enables using typescript discriminated unions for easy conditional typing).

# Implementation

- Port existing logic from content claims locator into the universal locator, initially based on the indexing service locator
   - this includes the cache
   - this also includes claimsFetched cache, though this has been modified to be a promise array, in order to resolve a potential data race that could cause two fetches for the same claim to be executed
- Write a content claims client that uses the existing content claims service plus some additional logic to mimic sending responses from the indexing service
- add additional tests to cover various scenarios not tested for the indexing service locator
- add locating content with the index (using a mocked indexer service running a very stripped down version of the indexer logic) to the fetcher spec, so that it now tests various scenarios for fetching content using both services.
- cover some additional edge cases not handled by the current content claims or indexer locator, such as location claims that have no range

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,13 +43,13 @@
       "import": "./src/fetcher/batching.js",
       "types": "./dist/src/fetcher/batching.d.ts"
     },
-    "./locator/content-claims": {
-      "import": "./src/locator/content-claims.js",
-      "types": "./dist/src/locator/content-claims.d.ts"
+    "./locator/content-claims-client": {
```
Comment (Peeja) [inline package.json:46]: **chore:** I'm not sure how we track versioning in this repo, but let's make sure this ends up as a major bump, since this is definitely a breaking change.

File: test/helpers/claims.js [hunk 2]
```diff
@@ -43,6 +51,111 @@ export const withClaimsServer = testfn =>
     }
   })
 
+/**
+ * @template {ClaimsServerContext} T
+ * @param {(assert: import('entail').Assert, ctx: T & IndexingServerContext) => unknown} testfn
```
Comment (Peeja) [inline test/helpers/claims.js:56]: **nit:** Let's use an `@import` rather than `import()`.

**Review by Peeja — COMMENTED** [review]
Some comments, but TBH, I think I've lost enough context (or rather the context has evolved) not to be confident in a full approval.

File: src/locator/content-claims-client.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      } else {
+        claims.push(...digestClaims)
+        for (const claim of digestClaims) {
+          if (claim.type === 'assert/index') {
+            this.#indexCids.set(claim.index.multihash, true)
+          }
+          if (claim.type === 'assert/location' && this.#indexCids.has(Claims.contentMultihash(claim))) {
+            try {
+              const fetchRes = await fetchIndex(claim)
+              indexBytes = await fetchRes.bytes()
```
Comment (alanshaw) [inline src/locator/content-claims-client.js:95]: break here?

File: src/locator/content-claims-client.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            } catch (err) {
+              console.warn('unable to fetch index', err instanceof Error ? err.message : 'unknown error')
```
Comment (alanshaw) [inline src/locator/content-claims-client.js:97]: "unknown error" is not useful, just log it and let console figure out formatting!
```suggestion
              console.warn('unable to fetch index', err)
```

File: src/locator/content-claims-client.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      return
+    }
+    if (this.#indexCids.has(digest)) {
+      const obj = await withSimpleSpan('carPark.get', this.#carpark.get, this.#carpark)(toBlobKey(digest))
+      if (!obj) {
+        return
+      }
+      indexBytes = new Uint8Array(await obj.arrayBuffer())
+    } else {
+      const obj = await this.#carpark.head(toBlobKey(digest))
```
Comment (alanshaw) [inline src/locator/content-claims-client.js:131]: You might want to also check if `[CARCID]/[CARCID].car` exists from store protocol days.

File: src/api.ts [hunk 6]
```diff
@@ -47,6 +47,10 @@ export interface ByteRange {
   length: number
 }
 
+type OptionalLengthByteRange = Partial<Pick<ByteRange, 'length'>> & Omit<ByteRange, 'length'>
+export type OptionalSite = { range?: OptionalLengthByteRange } & Omit<Site, 'range'>
+export type NoLengthLocation = { digest: MultihashDigest, site: OptionalSite[] }
```
[the remark this answers is not in the document]
Comment (alanshaw) [reply src/api.ts:52]: Having read more of the PR, what about calling this `ShardLocation` and `SliceLocation` where the former has the optional range?

File: src/locator/index.js [hunk 7]
```diff
@@ -1 +1,249 @@
...
+    return this.#cache.get(shardKey)
+  }
+
+  /**
+   *
+   * @param {API.MultihashDigest} digest
+   * @param {Kind} kind
+   */
+  async #executeReadClaims (digest, kind) {
+    const result = await this.#client.queryClaims({
```
Comment (alanshaw) [inline src/locator/index.js:150]: Am I correct in saying this can all be simplified once we switch to using the indexing service client instead of content claims?

**Review by alanshaw — APPROVED** [review]
I've left a few non-blocking comments but in general this LGTM

File: test/helpers/claims.js (hunk 2 above)
Comment (hannahhoward) [reply test/helpers/claims.js:56]: gonna do this but I do see it's spread all over the file.

File: src/locator/index.js (hunk 7 above)
Comment (hannahhoward) [reply src/locator/index.js:150]: not really unfortunately. only one part of executeReadClaims (looking for location claims for indexes) is safe to destroy.

File: src/locator/content-claims-client.js (hunk 5 above)
Comment (hannahhoward) [reply src/locator/content-claims-client.js:131]: so here's the reason why not:
- you will always get a location claim for any dag cid from content claims cause of the block index lookup. So it only applies to car/shard cids
- if it's actual car file retrieval at the root level, it will happen higher up the chain in freeway
- so the only scenario this code gets hit is for a CAR/shard from an index. That will never happen on store protocol cause no indexes.

File: src/locator/index.js [hunk 8]
```diff
@@ -1 +1,249 @@
...
+
+    // fetch location claims for any indexes we don't have a known shard for
+    for (const claim of result.ok.claims.values()) {
+      if (claim.type === 'assert/index') {
+        const location = this.#getShard(claim.index.multihash)
+        if (!location) {
+          await this.#readClaims(claim.index.multihash, 'location')
+        }
+      }
+    }
```
Comment (hannahhoward) [inline src/locator/index.js:185]: @alanshaw this is the bit we can remove when content claims goes away.

File: src/locator/index.js [hunk 9]
```diff
@@ -1 +1,249 @@
...
+      kind
+    })
+
+    if (result.error) return
+
+    // process any location claims
+    for (const claim of result.ok.claims.values()) {
+      if (claim.type === 'assert/location') {
+        if (claim.range?.length != null) {
+          addOrSetLocation(this.#cache, contentMultihash(claim), {
```
Comment (hannahhoward) [inline src/locator/index.js:162]: @alanshaw 

FWIW this is a complicated code path cause it applies to times:
1. Indexing service returns a stored location claim that has an range
2. Indexing service uses legacy block index to synthesize a location claim

That's why it's stored in cache -- it might be an actual block location you ultimately return to the top level locate path

File: src/api.ts [hunk 10]
```diff
@@ -47,6 +47,10 @@ export interface ByteRange {
   length: number
 }
 
+type OptionalLengthByteRange = Partial<Pick<ByteRange, 'length'>> & Omit<ByteRange, 'length'>
+export type OptionalSite = { range?: OptionalLengthByteRange } & Omit<Site, 'range'>
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply src/api.ts:51]: the avoid changing the public API? the blob-fetcher counts on these things not being null
